src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,186 @@
+@[simp] lemma map_add : X (f + f') = X f + X f' := derivation.map_add X f f'
+@[simp] lemma map_zero : X 0 = 0 := derivation.map_zero X
+@[simp] lemma map_neg : X (-f) = -X f := derivation.map_neg X f
+@[simp] lemma map_sub : X (f - f') = X f - X f' := derivation.map_sub X f f'
+
+instance : add_comm_group (left_invariant_derivation I G) :=
+coe_injective.add_comm_group _ coe_zero coe_add coe_neg coe_sub
+
+instance : has_scalar 𝕜 (left_invariant_derivation I G) :=
+{ smul := λ r X, ⟨r • X, λ f g, by { simp only [derivation.Rsmul_apply, algebra.id.smul_eq_mul,
+            mul_eq_mul_left_iff, linear_map.map_smul, left_invariant'], left, refl }⟩ }
+
+variables (r X)
+
+@[simp] lemma coe_smul : ⇑(r • X) = r • X := rfl
+@[simp] lemma map_smul : X (r • f) = r • X f := derivation.map_smul X r f
+@[simp] lemma leibniz : X (f * f') = f • X f' + f' • X f := X.leibniz' _ _
+@[simp] lemma lift_smul (k : 𝕜) : (↑(k • X) : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) = k • X := rfl
+
+variables (I G)
+
+/-- The coercion to function is a monoid homomorphism. -/
+def coe_fn_add_monoid_hom : (left_invariant_derivation I G) →+ _ :=
+⟨λ X, X.to_derivation.to_fun, coe_zero, coe_add⟩
+
+variables {I G}
+
+instance : module 𝕜 (left_invariant_derivation I G) :=
+coe_injective.module _ (coe_fn_add_monoid_hom I G) coe_smul
+
+/-- Evaluation at a point for left invariant derivation. Same thing as for generic global
+derivations (`derivation.eval_at`). -/
+def eval_at : (left_invariant_derivation I G) →ₗ[𝕜] (point_derivation I g) :=
+{ to_fun := λ X, derivation.eval_at g ↑X,
+  map_add' := λ X Y, rfl,
+  map_smul' := λ k X, rfl }
+
+lemma eval_at_apply : eval_at g X f = (X f) g := rfl
+
+@[simp] lemma eval_at_coe : derivation.eval_at g ↑X = eval_at g X := rfl
+
+lemma left_invariant : (𝒅(𝑳 I g)) (1 : G) (eval_at (1 : G) X) f = eval_at g X f :=
+(X.left_invariant'' f g)
+
+lemma eval_at_mul : eval_at (g * h) X f = (𝒅(𝑳 I g)) h (eval_at h X) f :=
+by rw [←left_invariant, L_mul, fdifferential_comp, apply_fdifferential, linear_map.comp_apply,
+  apply_fdifferential, left_invariant]
+
+lemma comp_L : (X f).comp (𝑳 I g) = X (f.comp (𝑳 I g)) :=
+by ext h; rw [times_cont_mdiff_map.comp_apply, L_apply, ←eval_at_apply, eval_at_mul,
+  apply_fdifferential, eval_at_apply]
+
+instance : has_bracket (left_invariant_derivation I G) (left_invariant_derivation I G) :=
+{ bracket := λ X Y, ⟨⁅(X : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯), Y⁆, λ f g, begin
+    have hX := left_invariant' g X (Y f), have hY := left_invariant' g Y (X f),
+    rw [apply_fdifferential, derivation.eval_apply] at hX hY ⊢, rw [comp_L] at hX hY,
+    rw [derivation.commutator_apply, smooth_map.coe_sub, pi.sub_apply, coe_derivation],
+    rw [coe_derivation] at hX hY ⊢, rw [hX, hY], refl end⟩ }
+
+@[simp] lemma commutator_coe_derivation :
+  ⇑⁅X, Y⁆ = (⁅(X : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯), Y⁆ :
+  derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) := rfl
+
+lemma commutator_apply : ⁅X, Y⁆ f = X (Y f) - Y (X f) := rfl
```

Comment:
I suspect this should use the word `lie` or `bracket` in the name rather than `commutator`. Maybe @ocfnash can comment.

Reply:
It was `commutator` one year ago when I wrote the code but something might have changed

Reply:
For what it's worth `lie_apply` is my favourite and `commutator_apply` is my least favourite, but I think all three options OK.

Reply:
Well, I see now that commutator is consistent with the derivation file! I don't care about the name but I care about being consistent with the derivation file, so I would leave this refactor to some other PR, which takes cares of making the names consistent across the whole library. It would have not much to do with this PR.